Parquet/src/main/java/io/deephaven/parquet/ParquetFileReader.java

```diff
@@ -77,52 +73,66 @@ public ParquetFileReader(String filePath, SeekableChannelsProvider channelsProvi
         type = fromParquetSchema(fileMetaData.schema,fileMetaData.column_orders);
     }
 
-    private HashSet<String> columnsWithDictionaryUsedOnEveryDataPage = null;
-    public HashSet<String> getColumnsWithDictionaryUsedOnEveryDataPage() {
+    private Set<String> columnsWithDictionaryUsedOnEveryDataPage = null;
+    public Set<String> getColumnsWithDictionaryUsedOnEveryDataPage() {
         if (columnsWithDictionaryUsedOnEveryDataPage == null) {
             columnsWithDictionaryUsedOnEveryDataPage = calculateColumnsWithDictionaryUsedOnEveryDataPage();
         }
         return columnsWithDictionaryUsedOnEveryDataPage;
     }
 
-    private HashSet<String> calculateColumnsWithDictionaryUsedOnEveryDataPage() {
+    /** True only if we are certain every data page in this column chunk uses dictionary encoding;
+     *  note false also covers the "we can't tell" case. */
+    private boolean columnChunkUsesDictionaryOnEveryPage(final ColumnChunk columnChunk) {
+        final ColumnMetaData columnMeta = columnChunk.getMeta_data();
+        if (columnMeta.encoding_stats == null) {
+            return false;  // this is false as "don't know".
+        }
+        for (PageEncodingStats encodingStat : columnMeta.encoding_stats) {
+            if (encodingStat.page_type != PageType.DATA_PAGE
+                    && encodingStat.page_type != PageType.DATA_PAGE_V2) {
+                // skip non-data pages.
+                continue;
+            }
+            // this is a data page.
+            if (encodingStat.encoding != Encoding.PLAIN_DICTIONARY
+                    && encodingStat.encoding != Encoding.RLE_DICTIONARY) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private Set<String> calculateColumnsWithDictionaryUsedOnEveryDataPage() {
         final HashSet<String> result = new HashSet<>(fileMetaData.getSchemaSize());
-        int rowGroupSeq = 0;
-        for (RowGroup rowGroup : fileMetaData.getRow_groups()) {
-            // On the first pass, for row group zero, we are going to add all columns to the set
-            // that satisfy the restriction.
-            // On later passes after zero, we will remove any column that does not satisfy
-            // the restriction.
+        final List<RowGroup> rowGroups = fileMetaData.getRow_groups();
+        final Iterator<RowGroup> riter = rowGroups.iterator();
+        if (!riter.hasNext()) {
+            return result;
```

Comment:
I think this should be the full map, not the empty map, maybe? Do we have a way to get such columns, without a row group? This is probably a meaningless distinction in the single partition case.

Reply:
We do have a way to get all column names without going through the chunks, we can just do:
```
            for (SchemaElement se : fileMetaData.getSchema()) {
                if (!se.isSetNum_children()) {  // We want only the leaves.
                    result.add(se.getName());
                }
            }
```
(I was thinking of doing it and wrote the code above).

But I am changing my mind.  I think if we define the property as:

```
    /**
     * Get the name of all columns that we can know for certain
     * (a) have a dictionary, and (b) use the dictionary on all data pages.
     *
     * @return A set of parquet column names that satisfies the required condition.
     */
    public Set<String> getColumnsWithDictionaryUsedOnEveryDataPage() {
    ...
```

This is properly empty, and I think that is a better result in terms of behavior (goes with the "when we can't tell")

Reply:
Changed my mind, you convinced me, now we return all columns.